server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -625,45 +630,70 @@ private static void validate(final String repositoryName) {
         for (SnapshotsInProgress.Entry snapshot : snapshots.entries()) {
             if (repository.equals(snapshot.snapshot().getRepository())) {
-                return true;
+                throw newRepositoryInUseException(repository, "snapshot is in progress");
             }
         }
         for (SnapshotDeletionsInProgress.Entry entry :
             clusterState.custom(SnapshotDeletionsInProgress.TYPE, SnapshotDeletionsInProgress.EMPTY).getEntries()) {
             if (entry.repository().equals(repository)) {
-                return true;
+                throw newRepositoryInUseException(repository, "snapshot deletion is in progress");
             }
         }
         for (RepositoryCleanupInProgress.Entry entry :
             clusterState.custom(RepositoryCleanupInProgress.TYPE, RepositoryCleanupInProgress.EMPTY).entries()) {
             if (entry.repository().equals(repository)) {
-                return true;
+                throw newRepositoryInUseException(repository, "repository clean up is in progress");
             }
         }
         for (RestoreInProgress.Entry entry : clusterState.custom(RestoreInProgress.TYPE, RestoreInProgress.EMPTY)) {
             if (repository.equals(entry.snapshot().getRepository())) {
-                return true;
+                throw newRepositoryInUseException(repository, "snapshot restore is in progress");
             }
         }
+    }
+
+    private static void ensureNoSearchableSnapshotsIndicesInUse(ClusterState clusterState, RepositoryMetadata repositoryMetadata) {
+        long count = 0L;
+        List<Index> indices = null;
+        for (IndexMetadata indexMetadata : clusterState.metadata()) {
+            if (indexSettingsMatchRepositoryMetadata(indexMetadata.getSettings(), repositoryMetadata)) {
+                if (indices == null) {
+                    indices = new ArrayList<>();
+                }
+                if (indices.size() < 5) {
+                    indices.add(indexMetadata.getIndex());
+                }
+                count += 1L;
+            }
+        }
+        if (indices != null && indices.isEmpty() == false) {
+            throw newRepositoryInUseException(repositoryMetadata.name(),"found " + count
+                + " searchable snapshots indices that use the repository: "
+                + Strings.collectionToCommaDelimitedString(indices)
+                + (count > indices.size() ? ",..." : "")
+            );
+        }
+    }
+
+    private static boolean indexSettingsMatchRepositoryMetadata(Settings indexSettings, RepositoryMetadata repositoryMetadata) {
+        if ("snapshot".equals(INDEX_STORE_TYPE_SETTING.get(indexSettings))) {
+            return Objects.equals(repositoryMetadata.uuid(), indexSettings.get("index.store.snapshot.repository_uuid"))
+                || Objects.equals(repositoryMetadata.name(), indexSettings.get("index.store.snapshot.repository_name"));
```

Comment:
If `indexSettings.get("index.store.snapshot.repository_uuid")` is nonempty then we shouldn't care about the repo name I think. E.g. in Cloud it should be ok to remove/adjust the `found-snapshots` repository if all your searchable snapshots are from a different cluster's `found-snapshots` repo which you're calling something different.

Reply:
It makes sense, thanks for spotting this. I pushed 700dffb738f3fee73c9349d7de94428c10965823